perf(ivy): move attributes array into component def

crisbeto opened the pull request:
Currently Ivy stores the element attributes into an array above the component def and passes it into the relevant instructions, however the problem is that upon minification the array will get a unique name which won't compress very well. These changes move the attributes array into the component def and pass in the index into the instructions instead.

Before:
```
const _c0 = ['foo', 'bar'];

SomeComp.ngComponentDef = defineComponent({
  template: function() {
    element(0, 'div', _c0);
  }
});
```

After:
```
SomeComp.ngComponentDef = defineComponent({
  consts: [['foo', 'bar']],
  template: function() {
    element(0, 'div', 0);
  }
});
```

A couple of cases that this PR doesn't handle:
* Template references are still in a separate array.
* i18n attributes are still in a separate array.

mhevery commented on packages/compiler-cli/test/compliance/r3_view_compiler_listener_spec.ts:
```diff
@@ -195,18 +193,18 @@ describe('compiler compliance: listen()', () => {
           selectors: [["my-component"]],
           consts: 4,
           vars: 0,
+          attrs: [[${AttributeMarker.Bindings}, "click"]],
```
Instead of `attrs` property, can this be name something more generic? This is because in the future we will store other things here such as sub-template functions, references, etc.

Options:
1. Name it `refs`/or something else.
2. Rename `consts` => `decls`, and then name this `consts`.

crisbeto replied:
I went with option 2 since `refs` could be mistaken for template references.

mhevery commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -1279,10 +1282,21 @@ export class TemplateDefinitionBuilder implements t.Visitor<void>, LocalResolver
-    return attrsExprs.length > 0 ?
-        this.constantPool.getConstLiteral(o.literalArr(attrsExprs), true) :
-        o.TYPED_NULL_EXPR;
+  private addAttributes(attrsExprs: o.Expression[]): o.LiteralExpr {
```
probably name these methods something more generic so that it is clear that we can store other things than attributes.

mhevery commented on packages/core/src/render3/definition.ts:
```diff
@@ -171,6 +172,9 @@ export function ɵɵdefineComponent<T>(componentDefinition: {
   template: ComponentTemplate<T>;
 
+  /** Attributes for all of the nodes inside the component template. Used during creation. */
+  attrs?: TAttributes[];
```
per discussion above this should be name something else.

crisbeto committed:
perf(ivy): move attributes array into component def

Currently Ivy stores the element attributes into an array above the component def and passes it into the relevant instructions, however the problem is that upon minification the array will get a unique name which won't compress very well. These changes move the attributes array into the component def and pass in the index into the instructions instead.

Before:
```
const _c0 = ['foo', 'bar'];

SomeComp.ngComponentDef = defineComponent({
  template: function() {
    element(0, 'div', _c0);
  }
});
```

After:
```
SomeComp.ngComponentDef = defineComponent({
  consts: [['foo', 'bar']],
  template: function() {
    element(0, 'div', 0);
  }
});
```

A couple of cases that this PR doesn't handle:
* Template references are still in a separate array.
* i18n attributes are still in a separate array.

crisbeto committed:
fixup! perf(ivy): move attributes array into component def

kara commented on packages/core/src/render3/instructions/element.ts:
```diff
@@ -43,24 +42,25 @@ import {registerInitialStylingOnTNode} from './styling';
   const lView = getLView();
   const tView = lView[TVIEW];
+  const tViewConsts = tView.consts;
+  const consts = tViewConsts === null || constsIndex == null ? null : tViewConsts[constsIndex];
```
Have we checked yet if there's any perf impact of the extra read using the micro-benchmarks?

Also, wouldn't `constIndex` be null if `tViewConsts` is null? Could we remove the `tViewConsts === null` check?

```suggestion
  const consts = constsIndex == null ? null : tViewConsts[constsIndex];
```

crisbeto replied:
I haven't run any if the benchmarks yet. Also the `constsIndex` could be null if we need to skip over it to pass in the `localRefs` param. I suppose we can get rid of it once the local refs are in the `consts` array.

kara commented:
Some stats: 

Before (left), this PR (right)

<img width="1680" alt="Screen Shot 2019-10-02 at 5 25 24 PM" src="https://user-images.githubusercontent.com/3871688/66096256-438cf880-e54f-11e9-8067-6a14d9d9c1fd.png">

Red squares -> yellow, showing some improvement

crisbeto committed:
fixup! perf(ivy): move attributes array into component def

crisbeto commented on packages/compiler-cli/test/compliance/r3_view_compiler_i18n_spec.ts:
```diff
@@ -269,32 +268,33 @@ describe('i18n support in the template compiler', () => {
             $r3$.ɵɵi18nAttributes(3, $_c5$);
             $r3$.ɵɵtext(4, "Content B");
             $r3$.ɵɵelementEnd();
-            $r3$.ɵɵelementStart(5, "div", $_c6$);
```
I don't think that variable was there to begin with, or at least it was folded under the `...`.

crisbeto committed:
fixup! perf(ivy): move attributes array into component def

petebacondarwin commented:
Would it be possible to make `consts` a function that is called once before the template function is first called? This would allow us to avoid running initialization code before it is needed.

It would be particularly useful for the `$localize` calls. Currently `$localize` is being called as soon as the component is imported into the application. This makes it problematic to find a suitable moment to load translations. Currently all these `I18N_...` constants get initialized so early it is really hard to ensure that runtime translations are loaded early enough.

crisbeto commented:
Makes sense to me @petebacondarwin. The question is whether we want to get these changes in and do a follow-up that turns it into a function and moves the i18n and references arrays into it.

mhevery commented:
> Makes sense to me @petebacondarwin. The question is whether we want to get these changes in and do a follow-up that turns it into a function and moves the i18n and references arrays into it.

I vote to land this and follow up later.

kara reviewed (approved):
LGTM, but this is still blocked on running the microbenchmarks

kara commented:
@crisbeto Thanks for doing that! Looks like this PR gives us a slight perf edge, so we're definitely good to merge.

alxhub commented:
[Presubmit](http://test/OCL:273587542:BASE:273587593:1570565028477:90157d73)
[Ivy Presubmit](http://test/OCL:273587542:BASE:273588062:1570565172833:19639556)

mhevery commented on packages/core/src/render3/interfaces/definition.ts:
```diff
@@ -241,6 +243,9 @@ export interface ComponentDef<T> extends DirectiveDef<T> {
   readonly template: ComponentTemplate<T>;
 
+  /** Constants associated with the component's view. */
+  readonly consts: TAttributes[]|null;
```
For follow up can we change this to:
```suggestion
  readonly consts: () => TAttributes[]|null;
```

mhevery reviewed (approved):
LGTM

Can we do a follow up PR:
- [ ] embedded template functions should also be moved to `consts`
- [ ] Also include localization strings in the `consts`.
- [ ] Change the `ComponentDef.const` to have a function which returns the arrays. The reason for this is that @ocombe and @petebacondarwin need a way to delay the execution of the `$localize` until after locale strings are loaded. By putting it into the closure this can be achieved. When `TView` is created and we copy the `consts` from `ComponentDef` to the `TView` we can execute the closure which will in turn execute the `$localize` and we would cache the resulting values in `TView`

kara commented:
merge-assistance: global approval
